test/testhelpers/ULPError.jl

```diff
@@ -0,0 +1,59 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+module ULPError
+    export ulp_error, ulp_error_maximum
+    @noinline function throw_invalid()
+        throw(ArgumentError("invalid"))
+    end
+    function ulp_error(accurate::AbstractFloat, approximate::AbstractFloat)
+        # the ULP error is usually not required to great accuracy, so `Float32` should be precise enough
+        zero_return = Float32(0)
+        inf_return = Float32(Inf)
+        # handle floating-point edge cases
+        accur_is_zero = iszero(accurate)
+        accur_is_nan = isnan(accurate)
+        accur_is_inf = isinf(accurate)
+        approx_is_inf = isinf(approximate)
+        approx_is_nan = isnan(approximate)
+        approx_is_zero = iszero(approximate)
+        if accur_is_nan || approx_is_nan
+            return if accur_is_nan === approx_is_nan
+                zero_return
+            else
+                inf_return
+            end
+        end
+        if approx_is_inf
+            return if accur_is_inf && (signbit(accurate) == signbit(approximate))
+                zero_return
+            else
+                inf_return
+            end
+        end
```

Comment:
maybe put all of this inside `!(isfinite(approx) && isfinite(accurate))`? Also, I don't think these variables are used enough to be worthwhile.